[Cloud] Gateways integration

dejanb opened the pull request.

lulf commented on active/004-providers-integration.md:
````diff
@@ -0,0 +1,270 @@
+* Provider tries to push data using `as` property to define exact device it's representing, like
+
+```bash
+echo '{"temp":42}' | http --auth 'gateway1@myapp:hey-rodney' POST https://http.sandbox.drogue.cloud/v1/foo?as=device1
````
Thinking about the current capabilities of the TTN integration, I'm not sure if its possible for the TTN http integration to pass the device id as a query parameter. Is there already a plan on how to handle that, having a builtin TTN payload transform that can read the device id from the payload? Would that just be something that is supported for certain providers? (I think drogue-cloud  already have that for TTN, @ctron ?)

dejanb replied:
@lulf I would take Hono's approach where each provider will have its own implementation

https://github.com/eclipse/hono/tree/master/adapters/lora-vertx/src/main/java/org/eclipse/hono/adapter/lora/providers

They define then path within the endpoint where they accept messages and format them accordingly.

lulf replied:
Does that mean that this example is only relevant for the IoT protocol gateway type of provider? Would we then surface the TTN provider by a special name like "ttn-service" as your device config alludes to?

dejanb replied:
> Does that mean that this example is only relevant for the IoT protocol gateway type of provider?
Yeah, this example is for the situations when we're in control of the urls and device id encodings.

> Would we then surface the TTN provider by a special name like "ttn-service" as your device config alludes to?

We already have `/ttn` exposed in our http endpoint for this purpose. I would assume we could add it for every supported provider that needs special mapping. 
So in this example, the `ttn-service` device is basically properly set TTN integration that can send in the name of `device1`.
It would be great if we could infer the integration point from the request and do this automatically. We can support that as well. I'm not sure if that would be applicable in all situations.

lulf replied:
I think it makes sense, I was initially confused if 'ttn-service' was a referring to a provider that you had to create in the same way as a iot protocol gateway, or if it was a hardcoded value one could use.

dejanb replied:
Great. I'll add some more explanations.

lulf reviewed (commented):
+1 to suggestions from chat. I have one comment.

lulf commented on active/004-providers-integration.md:
```diff
@@ -0,0 +1,198 @@
+    "spec": {
+        "providerSelector": {
+            "group": "my-gateways"
+        }
```
Should we allow both `providers` and `providerSelector`? Is the first one redundant now that we have selectors? 

Alternative syntax:

```
"selector": {
  "matchNames": ["gateway1", "gateway2"]
}
```

```
"selector": {
  "matchLabels": {
    "group": "my-gateways"
   }
}
```

dejanb replied:
I was taking a hint from here https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodename providing an easy way for I think will be 90% of use cases. I don't know if that's the best-practices or not, so I'm OK with changing the syntax.

lulf commented on active/004-providers-integration.md:
```diff
@@ -0,0 +1,198 @@
+    },
+    "spec": {
+        "providers": ["ttn-us-east-service", "ttn-us-west-service"],
+        "command-policy": "all"
```
Should probably make this `commandPolicy` since it seems we have standardized on camelCase elsewhere.

lulf reviewed (commented):
Nice work! It looks good overall, I just wanted to bikeshed about syntax :)

dejanb commented:
This is now updated according to the initial implementation and discussions we had. Not everything is implemented at the moment, but for the features that work, I added examples of how to configure it with `drg` client.

dejanb commented:
Merging this. We can make further changes later on.
